server/src/main/java/org/elasticsearch/common/lucene/index/ElasticsearchDirectoryReader.java

```diff
@@ -59,7 +64,23 @@ protected DirectoryReader doWrapDirectoryReader(DirectoryReader in) throws IOExc
      * @param shardId the shard ID to expose via the elasticsearch internal reader wrappers.
      */
     public static ElasticsearchDirectoryReader wrap(DirectoryReader reader, ShardId shardId) throws IOException {
-        return new ElasticsearchDirectoryReader(reader, new SubReaderWrapper(shardId), shardId);
+        return wrap(reader, shardId, null);
+    }
+
+    public static ElasticsearchDirectoryReader wrap(DirectoryReader reader, ShardId shardId, @Nullable ESCacheHelper esCacheHelper)
+        throws IOException {
+        return new ElasticsearchDirectoryReader(reader, new SubReaderWrapper(shardId), shardId, esCacheHelper);
+    }
+
+    /**
+     * Retrieves Elasticsearch's version of the reader cache helper (see {@link ESCacheHelper})
+     */
+    public static ESCacheHelper getESReaderCacheHelper(DirectoryReader reader) {
+        ElasticsearchDirectoryReader esReader = getElasticsearchDirectoryReader(reader);
+        if (esReader == null || esReader.esCacheHelper == null) {
```

Comment:
Maybe add a comment about the `esReader == null` case. My understanding is that it should never be null, but the leniency might be ok since it "just" means that the cache would be useless, and correctness would not be affected? Maybe we still want to assert about this case?

Reply:
It should indeed never be null. I've added an assertion to that effect (cada8802697), but kept things still a bit more lenient, as this wouldn't affect correctness).